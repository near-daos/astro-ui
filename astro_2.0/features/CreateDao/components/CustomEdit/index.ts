import CustomEdit from './CustomEdit';

export default CustomEdit;
