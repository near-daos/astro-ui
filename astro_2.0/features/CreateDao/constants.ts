export const STEPS = {
  INFO: 'info',
  KYC: 'kyc',
  LINKS: 'links',
  MEMBERS: 'members',
  PROPOSALS: 'proposals',
  ASSETS: 'assets',
};
