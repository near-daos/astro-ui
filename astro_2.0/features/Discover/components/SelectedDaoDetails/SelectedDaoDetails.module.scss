@import 'styles/typography';
@import 'styles/mixins';

.root {
  align-items: flex-start;
  display: flex;
  flex-direction: column;
  margin-top: 24px;
  width: 100%;

  @include tablet-landscape {
    flex-direction: row;
  }
}

.name {
  align-items: center;
  display: flex;
}

.flag {
  margin-right: 12px;
}

.details {
  margin-left: 12px;
  width: 265px;
}

.daoName {
  @extend %title3;
  @include ellipse-text;
  color: var(--color-black);
}

.address {
  @extend %caption1;
  align-items: center;
  color: var(--color-neutral-60);
  display: block;
  width: fit-content;
  word-break: break-all;
}

.addressId {
  margin-right: 6px;
}

.close {
  color: var(--color-neutral-60);
  margin-right: 12px;

  &:hover {
    color: var(--color-primary-50);
  }
}

.navButton {
  margin-top: 24px;
  text-transform: unset;
  width: 100%;

  @include tablet-landscape {
    margin-left: auto;
    margin-top: 6px;
    width: auto;
  }
}
