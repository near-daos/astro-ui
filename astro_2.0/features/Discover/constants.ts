import { Interval } from 'services/DaoStatsService/types';

export const CONTRACT = { contract: 'astro' };

export enum DaoStatsTopics {
  GENERAL_INFO = 'generalInfo',
  USERS_AND_ACTIVITY = 'usersAndActivity',
  GOVERNANCE = 'governance',
  FLOW = 'flow',
  TVL = 'tvl',
  TOKENS = 'tokens',
}

export enum GeneralInfoTabs {
  ACTIVITY = 'activity',
  ACTIVE_DAOS = 'activeDaos',
}

export enum UsersAndActivityTabs {
  ACTIVE_USERS = 'activeUsers',
  ALL_USERS_ON_PLATFORM = 'allUsersOnPlatform',
  USERS_MEMBERS_OF_DAO = 'usersMembersOfDao',
  AVERAGE_NUMBER_OF_USERS_PER_DAO = 'averageNumberOfUsersPerDao',
  NUMBER_OF_INTERACTIONS = 'numberOfInteractions',
  AVERAGE_NUMBER_OF_INTERACTIONS_PER_DAO = 'averageNumberOfInteractionsPerDao',
  ALL_USERS_PER_DAO = 'allUsersPerDao',
}

export enum GovernanceTabs {
  NUMBER_OF_PROPOSALS = 'numberOfProposals',
  VOTE_THROUGH_RATE = 'voteThroughRate',
  ACTIVE_PROPOSALS = 'activeProposals',
  ACTIVE_VOTES = 'activeVotes',
}

export enum FlowTabs {
  TOTAL_IN = 'totalIn',
  TOTAL_OUT = 'totalOut',
  INCOMING_TRANSACTIONS = 'incomingTransactions',
  OUTGOING_TRANSACTIONS = 'outgoingTransactions',
}

export enum TvlTabs {
  PLATFORM_TVL = 'platformTvl',
  VL_IN_BOUNTIES = 'vlInBounties',
  NUMBER_OF_BOUNTIES = 'numberOfBounties',
  VL_OF_BOUNTIES = 'vlOfBounties',
  TVL = 'tvl',
}

export enum TokensTabs {
  NUMBER_OF_FTS = 'numberOfFts',
  VL_OF_FTS = 'vlOfFts',
  NUMBER_OF_NFTS = 'numberOfNfts',
}

export const intervalOptions = [
  { label: 'daily', value: Interval.DAY },
  { label: 'weekly', value: Interval.WEEK },
  { label: 'monthly', value: Interval.MONTH },
];
