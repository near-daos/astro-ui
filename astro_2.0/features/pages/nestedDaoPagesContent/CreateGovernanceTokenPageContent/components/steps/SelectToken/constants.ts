export const TOKEN_OPTIONS = {
  NEW: 'NEW',
  EXISTING: 'EXISTING',
};
