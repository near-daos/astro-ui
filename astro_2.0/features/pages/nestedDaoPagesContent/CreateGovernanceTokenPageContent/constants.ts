export const STEPS = {
  INTRO: 'intro',
  SELECT_TOKEN: 'select-token',
  CREATE_TOKEN: 'create-token',
  WIZARD: 'wizard',
  CONTRACT_ACCEPTANCE: 'contract-acceptance',
  TOKEN_DISTRIBUTION: 'token-distribution',
  CHANGE_VOTING_POLICY: 'change-voting-policy',
};
