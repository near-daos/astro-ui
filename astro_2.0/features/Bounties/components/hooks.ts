import { useCallback, useRef, useState } from 'react';
import { useRouter } from 'next/router';
import { useAsyncFn, useLocalStorage, useLocation } from 'react-use';
import axios, { CancelTokenSource } from 'axios';

import { SputnikHttpService } from 'services/sputnik';
import { useModal } from 'components/modal';
import { ConfirmActionModal } from 'astro_2.0/components/ConfirmActionModal';

import { Bounty, BountyContext, BountyProposal } from 'types/bounties';
import { VoteAction } from 'types/proposal';

import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';

import { useWalletContext } from 'context/WalletContext';
import { PaginationResponse } from 'types/api';
import { VOTE_ACTION_SOURCE_PAGE } from 'constants/votingConstants';

export function useBountyControls(
  daoId: string,
  bounty?: Bounty
): {
  handleClaim: () => void;
  handleUnclaim: () => void;
  loading: boolean;
} {
  const router = useRouter();
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();
  const [loading, setLoading] = useState(false);

  const [showModal] = useModal(ConfirmActionModal);

  const onSuccessHandler = useCallback(async () => {
    router.reload();
  }, [router]);

  const handleClaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Claim',
    });

    if (res?.length && bounty) {
      setLoading(true);

      const dao = await SputnikHttpService.getDaoById(daoId);

      if (!dao) {
        return;
      }

      try {
        setVoteActionSource(pathname);

        await nearService?.claimBounty(daoId, {
          bountyId: bounty?.bountyId,
          deadline: bounty?.maxDeadline,
          bountyBond: dao.policy.bountyBond,
          gas: res[0],
          tokenId: bounty?.token,
        });

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    daoId,
    setVoteActionSource,
    pathname,
    nearService,
    onSuccessHandler,
  ]);

  const handleUnclaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Unclaim',
    });

    if (res?.length && bounty) {
      try {
        setLoading(true);
        setVoteActionSource(pathname);
        await nearService?.unclaimBounty(daoId, bounty?.bountyId);

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    setVoteActionSource,
    pathname,
    nearService,
    daoId,
    onSuccessHandler,
  ]);

  return {
    handleClaim,
    handleUnclaim,
    loading,
  };
}

export function useBountyVoting(
  daoId: string,
  proposal: BountyProposal
): {
  handleVote: (vote: VoteAction) => void;
  loading: boolean;
} {
  const router = useRouter();
  const [showModal] = useModal(ConfirmActionModal);
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();

  const [{ loading }, handleVote] = useAsyncFn(
    async (vote: VoteAction) => {
      const res = await showModal({
        title: 'Confirm Your Vote',
      });

      if (res?.length) {
        setVoteActionSource(pathname);
        await nearService?.vote(daoId, proposal.proposalId, vote, res[0]);
        await router.reload();
      }
    },
    [daoId, proposal, router, nearService, pathname, setVoteActionSource]
  );

  return {
    handleVote,
    loading,
  };
}

export function useBountySearch(): {
  handleSearch: (
    val: string
  ) => Promise<PaginationResponse<BountyContext[]> | null>;
  loading: boolean;
} {
  const router = useRouter();
  const daoId = router.query.dao as string;
  const { accountId } = useWalletContext();
  const cancelTokenRef = useRef<CancelTokenSource | null>(null);

  const [{ loading }, handleSearch] = useAsyncFn(async query => {
    if (cancelTokenRef.current) {
      cancelTokenRef.current?.cancel('Cancelled by new req');
    }

    const { CancelToken } = axios;
    const source = CancelToken.source();

    cancelTokenRef.current = source;

    return SputnikHttpService.findBountyContext({
      daoId: daoId || '',
      accountId,
      query,
      cancelToken: source.token,
    });
  }, []);

  return {
    handleSearch,
    loading,
  };
}

export function useHideBounty(): {
  handleSubmit: () => void;
  handleSelect: (id: string) => void;
  selected: string[];
  loading: boolean;
} {
  const router = useRouter();
  const showHidden = router.query?.bountyFilter === 'hidden';
  const daoId = router.query.dao as string;
  const { accountId, nearService, pkAndSignature } = useWalletContext();
  const [selected, setSelected] = useState<string[]>([]);

  const [{ loading }, handleSubmit] = useAsyncFn(async () => {
    if (!pkAndSignature) {
      return;
    }

    try {
      const { publicKey, signature } = pkAndSignature;

      if (publicKey && signature) {
        await SputnikHttpService.toggleBountyContexts({
          accountId,
          publicKey,
          signature,
          daoId,
          ids: selected,
          isArchived: !showHidden,
        });

        await router.replace(router.asPath, undefined, {
          shallow: false,
        });

        setSelected([]);
      }
    } catch (err) {
      showNotification({
        type: NOTIFICATION_TYPES.ERROR,
        lifetime: 20000,
        description: err.message,
      });
    }
  }, [
    selected,
    daoId,
    showHidden,
    accountId,
    router,
    nearService,
    pkAndSignature,
  ]);

  const handleSelect = useCallback(
    (id: string) => {
      if (selected.includes(id)) {
        setSelected(selected.filter(item => item !== id));
      } else {
        setSelected([...selected, id]);
      }
    },
    [selected]
  );

  return {
    handleSubmit,
    handleSelect,
    selected,
    loading,
  };
}
