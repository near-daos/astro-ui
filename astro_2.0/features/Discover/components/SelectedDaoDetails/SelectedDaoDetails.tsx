import React, { FC, useEffect, useState, useCallback } from 'react';
import { useMountedState } from 'react-use';
import { useRouter } from 'next/router';

import useQuery from 'hooks/useQuery';
import { DAO } from 'types/dao';

import { shortenString } from 'utils/format';
import { SINGLE_DAO_PAGE } from 'constants/routing';

import { SputnikHttpService } from 'services/sputnik';

import { Button } from 'components/button/Button';
import { CopyButton } from 'astro_2.0/components/CopyButton';
import { IconButton } from 'components/button/IconButton';
import { DaoLogo } from 'astro_2.0/features/DaoDashboardHeader/components/DaoLogo';

import styles from './SelectedDaoDetails.module.scss';

export const SelectedDaoDetails: FC = () => {
  const isMounted = useMountedState();
  const router = useRouter();
  const [data, setData] = useState<DAO | null>(null);
  const { query, updateQuery } = useQuery<{
    dao: string;
  }>({ shallow: true });

  useEffect(() => {
    if (query.dao) {
      SputnikHttpService.getDaoById(query.dao).then(res => {
        if (isMounted()) {
          setData(res);
        }
      });
    } else {
      setData(null);
    }
  }, [isMounted, query.dao]);

  const goToDaoPage = useCallback(() => {
    router.push({
      pathname: SINGLE_DAO_PAGE,
      query: {
        dao: data?.id,
      },
    });
  }, [data, router]);

  const onClose = useCallback(() => {
    updateQuery('dao', '');
  }, [updateQuery]);

  if (!data) {
    return null;
  }

  return (
    <div className={styles.root}>
      <IconButton
        onClick={onClose}
        icon="closeCircle"
        className={styles.close}
      />
      <div className={styles.name}>
        <DaoLogo src={data.flagCover || data.flagLogo || data.logo} size="md" />
        <div className={styles.details}>
          <div className={styles.daoName}>{data.name ?? data.id}</div>
          <div className={styles.address}>
            <CopyButton text={data.id} tooltipPlacement="auto">
              <div className={styles.addressId}>
                {shortenString(data.id, 36)}
              </div>
            </CopyButton>
          </div>
        </div>
      </div>
      <Button
        className={styles.navButton}
        size="medium"
        variant="secondary"
        onClick={goToDaoPage}
      >
        Go to DAO page
      </Button>
    </div>
  );
};
