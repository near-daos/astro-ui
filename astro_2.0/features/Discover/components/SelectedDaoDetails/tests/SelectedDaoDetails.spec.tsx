/* eslint-disable @typescript-eslint/ban-ts-comment */

import { useRouter } from 'next/router';
import { render } from 'jest/testUtils';
import { fireEvent } from '@testing-library/dom';
import { act, screen } from '@testing-library/react';

import { SINGLE_DAO_PAGE } from 'constants/routing';

import useQuery from 'hooks/useQuery';

import { SputnikHttpService } from 'services/sputnik';

import { SelectedDaoDetails } from 'astro_2.0/features/Discover/components/SelectedDaoDetails';

import { daoMock } from './mock';

jest.mock('hooks/useQuery', () => {
  return jest.fn(() => ({
    query: {},
  }));
});

jest.mock('next/router', () => {
  return {
    useRouter: jest.fn(),
  };
});

describe('SelectedDaoDetails', () => {
  it('Should render nothing if no data', () => {
    const { container } = render(<SelectedDaoDetails />);

    expect(container).toMatchSnapshot();
  });

  it('Should render component', async () => {
    // @ts-ignore
    useQuery.mockImplementation(() => ({
      query: {
        dao: '123',
      },
    }));

    jest
      .spyOn(SputnikHttpService, 'getDaoById')
      .mockImplementationOnce(() => Promise.resolve(daoMock));

    await act(async () => {
      render(<SelectedDaoDetails />);
    });

    expect(screen.getByText('Go to DAO page')).toBeInTheDocument();
  });

  it('Should navigate to dao page', async () => {
    // @ts-ignore
    useQuery.mockImplementation(() => ({
      query: {
        dao: '123',
      },
    }));

    const router = {
      push: jest.fn(),
    };

    // @ts-ignore
    useRouter.mockImplementation(() => router);

    jest
      .spyOn(SputnikHttpService, 'getDaoById')
      .mockImplementationOnce(() => Promise.resolve(daoMock));

    await act(async () => {
      render(<SelectedDaoDetails />);
    });

    fireEvent.click(screen.getAllByRole('button')[3]);

    expect(router.push).toBeCalledWith({
      pathname: SINGLE_DAO_PAGE,
      query: {
        dao: 'default-flag-test.sputnikv2.testnet',
      },
    });
  });

  it('Should properly handle close icon click', async () => {
    const updateQuery = jest.fn();

    // @ts-ignore
    useQuery.mockImplementation(() => ({
      query: {
        dao: '123',
      },
      updateQuery,
    }));

    jest
      .spyOn(SputnikHttpService, 'getDaoById')
      .mockImplementationOnce(() => Promise.resolve(daoMock));

    await act(async () => {
      render(<SelectedDaoDetails />);
    });

    fireEvent.click(screen.getAllByRole('button')[0]);

    expect(updateQuery).toBeCalledWith('dao', '');
  });
});
