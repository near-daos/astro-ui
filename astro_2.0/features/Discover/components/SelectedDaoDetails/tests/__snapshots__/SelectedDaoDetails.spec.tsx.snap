// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectedDaoDetails Should render nothing if no data 1`] = `<div />`;
